pinot-common/src/main/java/com/linkedin/pinot/common/lineage/SegmentMergeLineage.java

```diff
@@ -0,0 +1,333 @@
+      Map<String, List<String>> groupToSegmentsForLevel = new HashMap<>();
+      for (Map.Entry<String, String> groupEntry : entry.getValue().entrySet()) {
+        String groupId = groupEntry.getKey();
+        String segmentsString = groupEntry.getValue();
+        List<String> segments = Arrays.asList(segmentsString.split(SEGMENT_DELIMITER));
+        groupToSegmentsForLevel.put(groupId, new ArrayList<>(segments));
+      }
+      groupToSegmentsMap.put(level, groupToSegmentsForLevel);
+    }
+    return new SegmentMergeLineage(tableNameWithType, segmentGroupLineageMap, groupToSegmentsMap);
+  }
+
+  public ZNRecord toZNRecord() {
+    ZNRecord record = new ZNRecord(_tableNameWithType);
+    record.setListFields(_parentGroupToChildrenGroupsMap);
+    Map<String, Map<String, String>> groupToSegmentsMap = new HashMap<>();
+
+    for (Map.Entry<Integer, Map<String, List<String>>> entry : _levelToGroupToSegmentsMap.entrySet()) {
+      String key = LEVEL_KEY_PREFIX + entry.getKey();
+      Map<String, String> groupSegmentsForLevel = new HashMap<>();
+      for (Map.Entry<String, List<String>> groupEntry : entry.getValue().entrySet()) {
+        String groupId = groupEntry.getKey();
+        String segments = String.join(SEGMENT_DELIMITER, groupEntry.getValue());
+        groupSegmentsForLevel.put(groupId, segments);
+      }
+      groupToSegmentsMap.put(key, groupSegmentsForLevel);
+    }
+    record.setMapFields(groupToSegmentsMap);
+
+    return record;
+  }
+
+  /**
+   * Add segment merge lineage information
+   *
+   * @param groupId a group id
+   * @param currentGroupSegments a list of segments that belongs to the group
+   * @param childrenGroups a list of children groups that the current group covers
+   */
+  public void addSegmentGroup(String groupId, List<String> currentGroupSegments, List<String> childrenGroups)
+      throws InvalidConfigException {
+    // Get group level
+    Integer groupLevel = getGroupLevel(childrenGroups);
+
+    // Update group to segments map
+    Map<String, List<String>> groupToSegmentMap =
+        _levelToGroupToSegmentsMap.computeIfAbsent(groupLevel, k -> new HashMap<>());
+    if (groupToSegmentMap.containsKey(groupId)) {
+      throw new InvalidConfigException("Group id : " + groupId + " already exists.");
+    }
+    groupToSegmentMap.put(groupId, new ArrayList<>(currentGroupSegments));
+    _levelToGroupToSegmentsMap.put(groupLevel, groupToSegmentMap);
+
+    // Update segment group lineage map
+    if (groupLevel > DEFAULT_GROUP_LEVEL) {
+      if (_parentGroupToChildrenGroupsMap.containsKey(groupId)) {
+        throw new InvalidConfigException("Group id : " + groupId + " already exists.");
+      } else {
+        _parentGroupToChildrenGroupsMap.put(groupId, new ArrayList<>(childrenGroups));
+      }
+    }
+
+    LOGGER.info("New group has been added successfully to the segment lineage. (groupId: {}, currentGroupSegments: {}, "
```

Comment:
pls include table name in the log and all exceptions thrown

Reply:
added